File: src/kernel/x86.c
typedef unsigned int   uint;
typedef unsigned short ushort;
typedef unsigned char  uchar;

void
stosb(void *addr, int data, int cnt)
{
  asm volatile("cld; rep stosb" :
               "=D" (addr), "=c" (cnt) :
               "0" (addr), "1" (cnt), "a" (data) :
               "memory", "cc");
}

void
stosl(void *addr, int data, int cnt)
{
  asm volatile("cld; rep stosl" :
               "=D" (addr), "=c" (cnt) :
               "0" (addr), "1" (cnt), "a" (data) :
               "memory", "cc");
}

uchar
inb(ushort port)
{
  uchar data;

  asm volatile("in %1,%0" : "=a" (data) : "d" (port));
  return data;
}

void
outb(ushort port, uchar data)
{
  asm volatile("out %0,%1" : : "a" (data), "d" (port));
}

void
lcr3(uint val)
{
  asm volatile("movl %0,%%cr3" : : "r" (val));
}

File: src/kernel/x86.rs
// Not ideal yet, would like to use rusts inline asm with inlining here
extern {
    #[fast_ffi]
    pub fn stosl(addr: *mut (), data: int, cnt: int) -> ();
    #[fast_ffi]
    pub fn stosb(addr: *mut (), data: int, cnt: int) -> ();
    #[fast_ffi]
    pub fn inb(port:u16) -> u8;
    #[fast_ffi]
    pub fn outb(port:u16, data: u8) -> ();
    #[fast_ffi]
    #[inline(always)]
    pub fn lcr3(val:uint) -> ();
}


pub unsafe fn cli() {
  asm!( "cli"
       :
       :
       :
       : "volatile"
       );
}

